File: implementations/node-express/src/db/sqlite.ts
import type { Database } from "sqlite3";
import { FilterError, NotAllowedError, quoteIdent } from "../errors.js";
import type { Condition } from "../filter.js";
import {
  type ColumnInfo,
  type ColumnRef,
  type CommonValueEntry,
  type CountEntry,
  cellToJson,
  type DbSource,
  findExact,
  groupReferencedBy,
  opSql,
  opTakesValue,
  type QueryOpts,
  type ReferencedByEntry,
  type TableData,
  type TableInfo,
} from "./types.js";

/** SQLite's only schema for a single database file. */
const ONLY_SCHEMA = "main";

function checkSchema(schema: string | undefined): void {
  if (schema !== undefined && schema !== ONLY_SCHEMA) {
    throw new NotAllowedError(`schema "${schema}"`);
  }
}

function dbAll<T extends Record<string, unknown>>(db: Database, sql: string, params: unknown[] = []): Promise<T[]> {
  return new Promise((resolve, reject) => {
    db.all<T>(sql, params, (err, rows) => {
      if (err) reject(err);
      else resolve(rows);
    });
  });
}

/** sqlite3 interrupt runs from another event-loop turn; `settled` avoids interrupting the next query. */
function bounded<T>(db: Database, timeoutMs: number, fn: () => Promise<T>): Promise<T> {
  return new Promise((resolve, reject) => {
    let settled = false;
    const timer = setTimeout(() => {
      if (!settled) db.interrupt();
    }, timeoutMs);
    fn().then(
      (value) => {
        settled = true;
        clearTimeout(timer);
        resolve(value);
      },
      (err: unknown) => {
        settled = true;
        clearTimeout(timer);
        reject(err instanceof Error ? err : new Error(String(err)));
      },
    );
  });
}

async function allowedTables(db: Database, timeoutMs: number): Promise<string[]> {
  const rows = await bounded(db, timeoutMs, () =>
    dbAll<{ name: string }>(
      db,
      "select name from sqlite_master where type = 'table' and name not like 'sqlite\\_%' escape '\\' order by name",
    ),
  );
  return rows.map((r) => r.name);
}

// PRAGMA cannot bind table names; callers supply an allow-listed table
// (spec/protocol.md §6).
async function allowedColumns(db: Database, table: string, timeoutMs: number): Promise<string[]> {
  const quoted = quoteIdent(table);
  const rows = await bounded(db, timeoutMs, () =>
    dbAll<{ cid: number; name: string }>(db, `select cid, name from pragma_table_info(${quoted}) order by cid`),
  );
  return rows.map((r) => r.name);
}

// Composite FKs are omitted rather than risk mislabeling columns (spec/protocol.md §5.4.1).
async function keyMetadata(
  db: Database,
  table: string,
  timeoutMs: number,
): Promise<{ pkColumns: Set<string>; fkColumns: Map<string, ColumnRef> }> {
  const quoted = quoteIdent(table);
  const { cols, fks } = await bounded(db, timeoutMs, async () => {
    const cols = await dbAll<{ cid: number; name: string; pk: number }>(
      db,
      `select cid, name, pk from pragma_table_info(${quoted}) order by cid`,
    );
    // (id, seq, table, from, to) — `id` groups columns belonging to the
    // same constraint (composite FKs share an id). The quoted
    // "table"/"from"/"to" are pragma_foreign_key_list's own fixed output
    // column names (SQL keywords needing escape), not the caller-supplied
    // table.
    //
    // "REFERENCES parent" shorthand leaves "to" NULL; ppk resolves it by
    // position against the parent's PK, correlated per FK row since each
    // can reference a different parent — unlike referencedBy's single
    // bound parent (docs/adapter-decisions.md §5.4.1).
    const fks = await dbAll<{ id: number; seq: number; table: string; from: string; to: string }>(
      db,
      `select fkl.id, fkl.seq, fkl."table", fkl."from", coalesce(fkl."to", ppk.name, 'rowid') as "to"
       from pragma_foreign_key_list(${quoted}) fkl
       left join pragma_table_info(fkl."table") ppk on ppk.pk = fkl.seq + 1`,
    );
    return { cols, fks };
  });

  const pkColumns = new Set(cols.filter((c) => c.pk > 0).map((c) => c.name));

  const byConstraint = new Map<number, { from: string; refTable: string; to: string }[]>();
  for (const fk of fks) {
    const list = byConstraint.get(fk.id) ?? [];
    list.push({ from: fk.from, refTable: fk.table, to: fk.to });
    byConstraint.set(fk.id, list);
  }
  const fkColumns = new Map<string, ColumnRef>();
  for (const members of byConstraint.values()) {
    if (members.length !== 1) continue; // composite FK: omit entirely
    const { from, refTable, to } = members[0];
    // SQLite has no schema namespace (see ONLY_SCHEMA), so `schema` is
    // never set here.
    fkColumns.set(from, { table: refTable, column: to });
  }
  return { pkColumns, fkColumns };
}

/** SQLite filter SQL uses `?`, `CAST(... AS TEXT)`, and `LIKE` for `ILIKE`. */
function buildWhereClauseSqlite(conditions: Condition[], columnNames: string[]): { where: string; values: string[] } {
  if (conditions.length === 0) {
    return { where: "", values: [] };
  }

  const allowed = new Set(columnNames);
  const values: string[] = [];
  let clause = "";

  conditions.forEach((cond, i) => {
    if (!allowed.has(cond.column)) {
      throw new NotAllowedError(`column "${cond.column}"`);
    }
    const keyword = cond.op === "ILIKE" ? "LIKE" : opSql(cond.op);
    const cast = `CAST(${quoteIdent(cond.column)} AS TEXT)`;

    let inner: string;
    if (opTakesValue(cond.op)) {
      if (cond.value === undefined) {
        throw new FilterError(`op "${cond.op}" requires a value`);
      }
      values.push(cond.value);
      inner = `${cast} ${keyword} ?`;
    } else {
      inner = `${cast} ${keyword}`;
    }

    const wrapped = cond.not ? `(NOT (${inner}))` : `(${inner})`;
    if (i > 0) {
      if (cond.logic === undefined) {
        throw new FilterError(`condition ${i} is missing logic`);
      }
      clause += cond.logic === "OR" ? " OR " : " AND ";
    }
    clause += wrapped;
  });

  return { where: ` where ${clause}`, values };
}

/** SQLite implementation; see docs/adapter-decisions.md for protocol differences. */
export class SqliteSource implements DbSource {
  constructor(private readonly db: Database) {}

  // timeoutMs is part of DbSource's signature but unused here — no
  // catalog to query, SQLite always has exactly one schema.
  async listSchemas(): Promise<string[]> {
    return [ONLY_SCHEMA];
  }

  async listTables(schema: string | undefined, timeoutMs: number): Promise<TableInfo[]> {
    checkSchema(schema);
    const names = await allowedTables(this.db, timeoutMs);
    // No obj_description equivalent in SQLite — comments unsupported.
    return names.map((name) => ({ name }));
  }

  async tableCounts(schema: string | undefined, timeoutMs: number): Promise<CountEntry[]> {
    checkSchema(schema);
    const tables = await allowedTables(this.db, timeoutMs);
    // SQLite has no reltuples-equivalent catalog estimate; -1 is the
    // documented "no estimate" sentinel (spec/protocol.md §5.3) rather
    // than a per-table COUNT(*) scan.
    return tables.map((table) => ({ table, approx_rows: -1 }));
  }

  async queryTable(schema: string | undefined, table: string, opts: QueryOpts, timeoutMs: number): Promise<TableData> {
    checkSchema(schema);
    const tables = await allowedTables(this.db, timeoutMs);
    const realTable = findExact(tables, table);
    if (!realTable) {
      throw new NotAllowedError(`table "${table}"`);
    }

    const columnNames = await allowedColumns(this.db, realTable, timeoutMs);
    let sort: string | undefined;
    if (opts.sort !== undefined) {
      sort = findExact(columnNames, opts.sort);
      if (!sort) {
        throw new NotAllowedError(`column "${opts.sort}"`);
      }
    }

    const { where: whereClause, values: filterValues } = buildWhereClauseSqlite(opts.filter, columnNames);
    const { pkColumns, fkColumns } = await keyMetadata(this.db, realTable, timeoutMs);

    const quotedTable = quoteIdent(realTable);
    const columnTypes = await bounded(this.db, timeoutMs, () =>
      dbAll<{ cid: number; name: string; type: string }>(
        this.db,
        `select cid, name, type from pragma_table_info(${quotedTable}) order by cid`,
      ),
    );

    const columns: ColumnInfo[] = columnTypes.map((ct) => {
      const col: ColumnInfo = {
        name: ct.name,
        // SQLite's declared column types can be empty (""); fall back to
        // a stable label rather than emitting "".
        type: ct.type && ct.type.length > 0 ? ct.type : "unknown",
      };
      if (pkColumns.has(ct.name)) {
        col.key = "pk";
        if (fkColumns.has(ct.name)) col.references = fkColumns.get(ct.name);
      } else if (fkColumns.has(ct.name)) {
        col.key = "fk";
        col.references = fkColumns.get(ct.name);
      }
      return col;
    });

    // Aliased back to the real column name: an un-aliased CAST(...)
    // expression's result-set label is the literal expression text on
    // SQLite (confirmed empirically), not the source column name — row
    // access by name below (cellToJson via col.name) would otherwise
    // silently read undefined for every cell.
    const selectList = columns.map((c) => `CAST(${quoteIdent(c.name)} AS TEXT) AS ${quoteIdent(c.name)}`).join(", ");
    const orderClause =
      sort !== undefined ? ` order by ${quotedTable}.${quoteIdent(sort)} ${opts.descending ? "desc" : "asc"}` : "";
    const sql = `select ${selectList} from ${quotedTable}${whereClause}${orderClause} limit ? offset ?`;
    const params = [...filterValues, opts.limit, opts.offset];

    const rows = await bounded(this.db, timeoutMs, () => dbAll<Record<string, unknown>>(this.db, sql, params));
    const outRows: Record<string, string | null>[] = rows.map((row) => {
      const out: Record<string, string | null> = {};
      for (const col of columns) {
        out[col.name] = cellToJson(row[col.name]);
      }
      return out;
    });

    return {
      columns,
      rows: outRows,
      // No reltuples-equivalent estimate to read; -1 is the documented
      // "no estimate" sentinel (spec/protocol.md §5.4.4), not a second
      // COUNT(*) scan on every page load.
      total_approx: -1,
    };
  }

  async commonValues(
    schema: string | undefined,
    table: string,
    column: string,
    timeoutMs: number,
  ): Promise<CommonValueEntry[]> {
    checkSchema(schema);
    const tables = await allowedTables(this.db, timeoutMs);
    const realTable = findExact(tables, table);
    if (!realTable) {
      throw new NotAllowedError(`table "${table}"`);
    }
    const columnNames = await allowedColumns(this.db, realTable, timeoutMs);
    if (!findExact(columnNames, column)) {
      throw new NotAllowedError(`column "${column}"`);
    }
    // No pg_stats equivalent to read; an empty list is the documented "no
    // statistics available" answer (spec/protocol.md §5.5), not a live
    // GROUP BY scan.
    return [];
  }

  async referencedBy(schema: string | undefined, table: string, timeoutMs: number): Promise<ReferencedByEntry[]> {
    checkSchema(schema);
    const tables = await allowedTables(this.db, timeoutMs);
    if (!findExact(tables, table)) {
      throw new NotAllowedError(`table "${table}"`);
    }

    // No reverse-FK index and no information_schema: walk every table's
    // pragma_foreign_key_list (the schema is parsed in memory on open).
    // The TVF argument m.name is a column reference — not spliced;
    // fkl."table" is bound. COLLATE NOCASE because the pragma returns the
    // referenced name as written in the DDL, which SQLite resolves
    // case-insensitively. The sqlite_master predicate here matches
    // allowedTables(), so every referrer is already allow-listed.
    //
    // "REFERENCES parent" with no parenthesised column list is valid DDL
    // meaning "parent's own primary key", and for that form the pragma
    // reports "to" as NULL rather than filling in the PK name. ppk resolves
    // it: pragma_table_info(table)'s pk column is the 1-indexed position of
    // a column within the parent's primary key, matching fkl.seq's
    // 0-indexed position within the FK's column list — SQLite requires an
    // omitted column list to align position-for-position with the parent
    // PK, so this pairing holds for composite keys too. The 'rowid'
    // fallback covers a parent with no declared primary key, where
    // SQLite's parent key is the implicit rowid.
    const rows = await bounded(this.db, timeoutMs, () =>
      dbAll<{ name: string; id: number; from: string; to: string }>(
        this.db,
        `select m.name, fkl.id, fkl."from", coalesce(fkl."to", ppk.name, 'rowid') as "to"
         from sqlite_master m
         join pragma_foreign_key_list(m.name) fkl
         left join pragma_table_info(?) ppk on ppk.pk = fkl.seq + 1
         where m.type = 'table'
           and m.name not like 'sqlite\\_%' escape '\\'
           and fkl."table" = ? collate nocase
         order by m.name, fkl.id, fkl.seq`,
        [table, table],
      ),
    );

    // SQLite FKs are unnamed; fk_<id> is a per-table-stable synthetic
    // label (spec/protocol.md §5.9 permits this).
    return groupReferencedBy(
      rows.map((r) => ({
        table: r.name,
        constraint: `fk_${r.id}`,
        columns: [{ from: r.from, to: r.to }],
      })),
    );
  }
}

File: implementations/go-nethttp/sqlite.go
//go:build sqlite

package ashurbanipal

import (
	"context"
	"database/sql"
	"fmt"
	"time"
)

// onlySchema is the only name ListSchemas ever returns — SQLite has no
// schema namespace above a single database file, mirroring how a bare
// ATTACH-less connection exposes its one implicit "main" schema
// (implementations/rust/core/src/db/sqlite.rs::ONLY_SCHEMA).
const onlySchema = "main"

// checkSchema rejects any requested schema other than onlySchema with the
// same NotAllowedError shape Postgres returns for a schema absent from its
// live allow-list, so callers don't need to special-case which backend
// rejected it.
func checkSchema(schema *string) error {
	if schema == nil || *schema == onlySchema {
		return nil
	}
	return &NotAllowedError{What: fmt.Sprintf("schema %q", *schema)}
}

// SQLiteSource is gated behind the `sqlite` build tag (opt-in, mirroring
// the Rust reference's `sqlite` Cargo feature) — see
// docs/adapter-decisions.md for the per-clause decisions this makes where
// Postgres-specific catalog/stats mechanisms have no SQLite equivalent.
//
// Unlike PostgresSource, catalog lookups here don't pin one *sql.Tx per
// operation: SQLite has no per-session state (no search_path/USE-database
// analog) that could drift between pooled connections, so the
// cross-connection consistency problem resolveSchema's transaction-pinning
// solves on Postgres/MySQL simply doesn't exist here — this matches the
// Rust reference's own choice (sqlite.rs's query_table acquires a fresh
// pooled connection per catalog call, not one held for the whole
// operation).
type SQLiteSource struct {
	db      *sql.DB
	timeout time.Duration
}

// NewSQLiteSource builds a DbSource backed by db, bounding every query
// (catalog and data alike) by queryTimeoutSecs. db must already be opened
// with a SQLite driver (e.g. modernc.org/sqlite).
func NewSQLiteSource(db *sql.DB, queryTimeoutSecs int) *SQLiteSource {
	return &SQLiteSource{db: db, timeout: time.Duration(queryTimeoutSecs) * time.Second}
}

var _ DbSource = (*SQLiteSource)(nil)

// bounded wraps every query in a context deadline. Verified empirically
// (not inferred from documentation, per this port's brief) against a real
// SQLite file with a single-connection pool: a slow recursive-CTE query
// canceled by ctx returns in ~1s (not the tens of seconds the full scan
// would take), and a query issued immediately afterward on the same
// physical connection completes without delay — proving modernc.org/sqlite
// actually aborts execution on cancellation rather than just abandoning
// the wait while the query keeps running server-side. Unlike the Rust
// reference (whose sqlx driver needed sqlite3_progress_handler because
// context cancellation there only stopped waiting, not the blocking C
// call), plain database/sql context cancellation is sufficient here — and
// unlike that progress handler, nothing needs clearing afterward.
func (c *SQLiteSource) bounded(ctx context.Context) (context.Context, context.CancelFunc) {
	return context.WithTimeout(ctx, c.timeout)
}

func (c *SQLiteSource) allowedTables(ctx context.Context) ([]string, error) {
	ctx, cancel := c.bounded(ctx)
	defer cancel()
	rows, err := c.db.QueryContext(ctx,
		`select name from sqlite_master
		 where type = 'table' and name not like 'sqlite\_%' escape '\'
		 order by name`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []string
	for rows.Next() {
		var name string
		if err := rows.Scan(&name); err != nil {
			return nil, err
		}
		out = append(out, name)
	}
	return out, rows.Err()
}

// allowedColumns' table argument must already be validated against
// allowedTables — pragma_table_info is a table-valued function that takes
// its argument as spliced SQL text, not a bound parameter, so this is the
// one identifier per query that's escaped rather than bound (mirrors
// sqlite.rs::allowed_columns).
func (c *SQLiteSource) allowedColumns(ctx context.Context, table string) ([]string, error) {
	ctx, cancel := c.bounded(ctx)
	defer cancel()
	rows, err := c.db.QueryContext(ctx,
		fmt.Sprintf("select cid, name from pragma_table_info(%s) order by cid", quoteIdent(table)))
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []string
	for rows.Next() {
		var cid int64
		var name string
		if err := rows.Scan(&cid, &name); err != nil {
			return nil, err
		}
		out = append(out, name)
	}
	return out, rows.Err()
}

// keyMetadata mirrors postgres.go's keyMetadata: composite FKs are dropped
// entirely rather than risk mislabeling which referencing column pairs
// with which referenced column (spec/protocol.md §5.4.1).
func (c *SQLiteSource) keyMetadata(ctx context.Context, table string) (map[string]bool, map[string]ColumnRef, error) {
	quoted := quoteIdent(table)
	ctx, cancel := c.bounded(ctx)
	defer cancel()

	pkRows, err := c.db.QueryContext(ctx,
		fmt.Sprintf("select name, pk from pragma_table_info(%s)", quoted))
	if err != nil {
		return nil, nil, err
	}
	pkColumns := map[string]bool{}
	for pkRows.Next() {
		var name string
		var pk int64
		if err := pkRows.Scan(&name, &pk); err != nil {
			pkRows.Close()
			return nil, nil, err
		}
		if pk > 0 {
			pkColumns[name] = true
		}
	}
	if err := pkRows.Err(); err != nil {
		pkRows.Close()
		return nil, nil, err
	}
	pkRows.Close()

	// (id, seq, table, from, to) — id groups columns belonging to the same
	// constraint (composite FKs share an id). The quoted "table"/"from"/"to"
	// are pragma_foreign_key_list's own fixed output column names (SQL
	// keywords needing escape), not the caller-supplied table.
	fkRows, err := c.db.QueryContext(ctx,
		fmt.Sprintf(`select id, "table", "from", "to" from pragma_foreign_key_list(%s)`, quoted))
	if err != nil {
		return nil, nil, err
	}
	type fkRow struct {
		id                 int64
		refTable, from, to string
	}
	byConstraint := map[int64][]fkRow{}
	for fkRows.Next() {
		var r fkRow
		if err := fkRows.Scan(&r.id, &r.refTable, &r.from, &r.to); err != nil {
			fkRows.Close()
			return nil, nil, err
		}
		byConstraint[r.id] = append(byConstraint[r.id], r)
	}
	if err := fkRows.Err(); err != nil {
		fkRows.Close()
		return nil, nil, err
	}
	fkRows.Close()

	fkColumns := map[string]ColumnRef{}
	for _, members := range byConstraint {
		if len(members) != 1 {
			continue // composite FK: omit entirely
		}
		m := members[0]
		fkColumns[m.from] = ColumnRef{Table: m.refTable, Column: m.to}
	}
	return pkColumns, fkColumns, nil
}

// sqliteBuildWhereClause is the SQLite equivalent of postgres.go's
// BuildWhereClause: `?` placeholders instead of `$N`, `CAST(col AS TEXT)`
// instead of `col::text`, and ILIKE mapped to plain LIKE — SQLite's LIKE
// is already ASCII case-insensitive by default, so there's no separate
// case-insensitive keyword to map to (docs/adapter-decisions.md §5.4.2).
func sqliteBuildWhereClause(conditions []Condition, columnNames []string) (string, []string, error) {
	if len(conditions) == 0 {
		return "", nil, nil
	}
	allowed := make(map[string]bool, len(columnNames))
	for _, c := range columnNames {
		allowed[c] = true
	}

	var values []string
	var clause []byte
	for i, cond := range conditions {
		if !allowed[cond.Column] {
			return "", nil, &NotAllowedError{What: fmt.Sprintf("column %q", cond.Column)}
		}
		if !validOps[cond.Op] {
			return "", nil, filterErr("condition %d has invalid op %q", i, cond.Op)
		}
		keyword := cond.Op
		if keyword == "ILIKE" {
			keyword = "LIKE"
		}
		cast := fmt.Sprintf("CAST(%s AS TEXT)", quoteIdent(cond.Column))
		var inner string
		if opTakesValue(cond.Op) {
			if cond.Value == nil {
				return "", nil, filterErr("op %q requires a value", cond.Op)
			}
			inner = fmt.Sprintf("%s %s ?", cast, keyword)
			values = append(values, *cond.Value)
		} else {
			inner = fmt.Sprintf("%s %s", cast, keyword)
		}
		wrapped := "(" + inner + ")"
		if cond.Not {
			wrapped = "(NOT (" + inner + "))"
		}
		if i > 0 {
			if cond.Logic == nil {
				return "", nil, filterErr("condition %d is missing logic", i)
			}
			if *cond.Logic == "OR" {
				clause = append(clause, " OR "...)
			} else {
				clause = append(clause, " AND "...)
			}
		}
		clause = append(clause, wrapped...)
	}
	return " where " + string(clause), values, nil
}

func (c *SQLiteSource) ListSchemas(ctx context.Context) ([]string, error) {
	return []string{onlySchema}, nil
}

func (c *SQLiteSource) ListTables(ctx context.Context, schema *string) ([]TableInfo, error) {
	if err := checkSchema(schema); err != nil {
		return nil, err
	}
	names, err := c.allowedTables(ctx)
	if err != nil {
		return nil, err
	}
	// No obj_description equivalent in SQLite — comments unsupported.
	out := make([]TableInfo, len(names))
	for i, name := range names {
		out[i] = TableInfo{Name: name}
	}
	return out, nil
}

func (c *SQLiteSource) TableCounts(ctx context.Context, schema *string) ([]CountEntry, error) {
	if err := checkSchema(schema); err != nil {
		return nil, err
	}
	names, err := c.allowedTables(ctx)
	if err != nil {
		return nil, err
	}
	// SQLite has no reltuples-equivalent catalog estimate; -1 is the
	// documented "no estimate" sentinel (spec/protocol.md §5.3) rather
	// than a per-table COUNT(*) scan. See docs/adapter-decisions.md.
	out := make([]CountEntry, len(names))
	for i, name := range names {
		out[i] = CountEntry{Table: name, ApproxRows: -1}
	}
	return out, nil
}

func (c *SQLiteSource) QueryTable(ctx context.Context, schema *string, table string, opts QueryOpts) (TableData, error) {
	if err := checkSchema(schema); err != nil {
		return TableData{}, err
	}
	tables, err := c.allowedTables(ctx)
	if err != nil {
		return TableData{}, err
	}
	realTable, ok := findExact(tables, table)
	if !ok {
		return TableData{}, &NotAllowedError{What: fmt.Sprintf("table %q", table)}
	}

	columnNames, err := c.allowedColumns(ctx, realTable)
	if err != nil {
		return TableData{}, err
	}
	var sort *string
	if opts.Sort != nil {
		found, ok := findExact(columnNames, *opts.Sort)
		if !ok {
			return TableData{}, &NotAllowedError{What: fmt.Sprintf("column %q", *opts.Sort)}
		}
		sort = &found
	}

	var whereClause string
	var filterValues []string
	if len(opts.Filter) > 0 {
		whereClause, filterValues, err = sqliteBuildWhereClause(opts.Filter, columnNames)
		if err != nil {
			return TableData{}, err
		}
	}

	pkColumns, fkColumns, err := c.keyMetadata(ctx, realTable)
	if err != nil {
		return TableData{}, err
	}

	typeCtx, typeCancel := c.bounded(ctx)
	typeRows, err := c.db.QueryContext(typeCtx,
		fmt.Sprintf("select cid, name, type from pragma_table_info(%s) order by cid", quoteIdent(realTable)))
	if err != nil {
		typeCancel()
		return TableData{}, err
	}
	type colType struct{ name, typ string }
	var columnTypes []colType
	for typeRows.Next() {
		var cid int64
		var ct colType
		if err := typeRows.Scan(&cid, &ct.name, &ct.typ); err != nil {
			typeRows.Close()
			typeCancel()
			return TableData{}, err
		}
		columnTypes = append(columnTypes, ct)
	}
	if err := typeRows.Err(); err != nil {
		typeRows.Close()
		typeCancel()
		return TableData{}, err
	}
	typeRows.Close()
	typeCancel()

	columns := make([]ColumnInfo, 0, len(columnTypes))
	for _, ct := range columnTypes {
		col := ColumnInfo{Name: ct.name, Type: ct.typ}
		// SQLite's declared column types can be empty (dynamic typing);
		// fall back to a stable label rather than emitting "".
		if col.Type == "" {
			col.Type = "unknown"
		}
		if ref, ok := fkColumns[ct.name]; ok {
			r := ref
			col.References = &r
		}
		switch {
		case pkColumns[ct.name]:
			col.Key = KeyPK
		case col.References != nil:
			col.Key = KeyFK
		}
		columns = append(columns, col)
	}

	selectParts := make([]string, len(columns))
	for i, col := range columns {
		selectParts[i] = fmt.Sprintf("CAST(%s AS TEXT)", quoteIdent(col.Name))
	}
	selectList := joinComma(selectParts)

	orderClause := ""
	if sort != nil {
		direction := "asc"
		if opts.Descending {
			direction = "desc"
		}
		orderClause = fmt.Sprintf(" order by %s.%s %s", quoteIdent(realTable), quoteIdent(*sort), direction)
	}

	// Identifiers spliced here are schema-validated (realTable via
	// allowedTables, columns via allowedColumns, sort via the findExact
	// check above, filter columns via sqliteBuildWhereClause's own
	// allow-list check); every value is a bound ? parameter.
	query := fmt.Sprintf("select %s from %s%s%s limit ? offset ?",
		selectList, quoteIdent(realTable), whereClause, orderClause)

	args := make([]interface{}, 0, len(filterValues)+2)
	for _, v := range filterValues {
		args = append(args, v)
	}
	args = append(args, opts.Limit, opts.Offset)

	queryCtx, queryCancel := c.bounded(ctx)
	defer queryCancel()
	rows, err := c.db.QueryContext(queryCtx, query, args...)
	if err != nil {
		return TableData{}, err
	}
	defer rows.Close()

	var out []map[string]*string
	for rows.Next() {
		cells := make([]interface{}, len(columns))
		values := make([]cellValue, len(columns))
		for i := range values {
			cells[i] = &values[i]
		}
		if err := rows.Scan(cells...); err != nil {
			return TableData{}, err
		}
		row := make(map[string]*string, len(columns))
		for i, col := range columns {
			row[col.Name] = values[i].asJSON()
		}
		out = append(out, row)
	}
	if err := rows.Err(); err != nil {
		return TableData{}, err
	}
	if out == nil {
		out = []map[string]*string{}
	}

	return TableData{
		Columns: columns,
		Rows:    out,
		// No reltuples-equivalent estimate to read; -1 is the documented
		// "no estimate" sentinel (spec/protocol.md §5.4.4), not a live
		// COUNT(*) scan.
		TotalApprox: -1,
	}, nil
}

func (c *SQLiteSource) CommonValues(ctx context.Context, schema *string, table, column string) ([]CommonValueEntry, error) {
	if err := checkSchema(schema); err != nil {
		return nil, err
	}
	tables, err := c.allowedTables(ctx)
	if err != nil {
		return nil, err
	}
	realTable, ok := findExact(tables, table)
	if !ok {
		return nil, &NotAllowedError{What: fmt.Sprintf("table %q", table)}
	}
	columns, err := c.allowedColumns(ctx, realTable)
	if err != nil {
		return nil, err
	}
	if _, ok := findExact(columns, column); !ok {
		return nil, &NotAllowedError{What: fmt.Sprintf("column %q", column)}
	}
	// No pg_stats equivalent to read; an empty list is the documented "no
	// statistics available" answer (spec/protocol.md §5.5), not a live
	// GROUP BY scan. See docs/adapter-decisions.md.
	return []CommonValueEntry{}, nil
}
